File: helix-runtime/apps/android/app/src/main/java/com/helix/email/EmailService.kt
package com.helix.email

import android.util.Log
import com.helix.gateway.GatewaySession
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlin.coroutines.suspendCancellableCoroutine

class EmailService(private val gateway: GatewaySession) {
    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    // MARK: - Email Operations

    suspend fun fetchEmails(
        accountId: String,
        limit: Int = 50,
        offset: Int = 0,
    ): List<Email> = suspendCancellableCoroutine { continuation ->
        try {
            gateway.request(
                method = "email.list",
                params = mapOf(
                    "account_id" to accountId,
                    "limit" to limit,
                    "offset" to offset,
                ),
            ) { response ->
                try {
                    val emails = (response as? JsonElement)?.jsonArray?.map {
                        json.decodeFromJsonElement(Email.serializer(), it)
                    } ?: emptyList()
                    continuation.resume(emails)
                } catch (e: Exception) {
                    continuation.resumeWithException(EmailError.DecodingError(e.message ?: "Unknown"))
                }
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    suspend fun getEmail(id: String): Email = suspendCancellableCoroutine { continuation ->
        try {
            gateway.request(
                method = "email.get",
                params = mapOf("id" to id),
            ) { response ->
                try {
                    val email = json.decodeFromJsonElement(
                        Email.serializer(),
                        response as JsonElement
                    )
                    continuation.resume(email)
                } catch (e: Exception) {
                    continuation.resumeWithException(EmailError.DecodingError(e.message ?: "Unknown"))
                }
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    suspend fun searchEmails(filter: EmailSearchFilter, limit: Int = 50): List<Email> =
        suspendCancellableCoroutine { continuation ->
            try {
                val params = mutableMapOf<String, Any>(
                    "query" to filter.query,
                    "limit" to limit,
                )

                filter.from?.let { params["from"] = it }
                filter.to?.let { params["to"] = it }
                filter.beforeDate?.let { params["before_date"] = it }
                filter.afterDate?.let { params["after_date"] = it }
                filter.hasAttachment?.let { params["has_attachment"] = it }
                filter.isRead?.let { params["is_read"] = it }
                filter.isStarred?.let { params["is_starred"] = it }
                if (filter.labels.isNotEmpty()) params["labels"] = filter.labels

                gateway.request(
                    method = "email.search",
                    params = params,
                ) { response ->
                    try {
                        val emails = (response as? JsonElement)?.jsonArray?.map {
                            json.decodeFromJsonElement(Email.serializer(), it)
                        } ?: emptyList()
                        continuation.resume(emails)
                    } catch (e: Exception) {
                        continuation.resumeWithException(
                            EmailError.DecodingError(e.message ?: "Unknown")
                        )
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    suspend fun markAsRead(id: String, isRead: Boolean = true): Unit =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "email.mark_read",
                    params = mapOf("id" to id, "is_read" to isRead),
                ) { _ ->
                    continuation.resume(Unit)
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    suspend fun markAsStarred(id: String, isStarred: Boolean = true): Unit =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "email.mark_starred",
                    params = mapOf("id" to id, "is_starred" to isStarred),
                ) { _ ->
                    continuation.resume(Unit)
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    suspend fun addLabel(id: String, label: String): Unit =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "email.add_label",
                    params = mapOf("id" to id, "label" to label),
                ) { _ ->
                    continuation.resume(Unit)
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    suspend fun removeLabel(id: String, label: String): Unit =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "email.remove_label",
                    params = mapOf("id" to id, "label" to label),
                ) { _ ->
                    continuation.resume(Unit)
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    suspend fun deleteEmail(id: String): Unit = suspendCancellableCoroutine { continuation ->
        try {
            gateway.request(
                method = "email.delete",
                params = mapOf("id" to id),
            ) { _ ->
                continuation.resume(Unit)
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    suspend fun sendEmail(
        accountId: String,
        to: List<String>,
        subject: String,
        body: String,
        cc: List<String> = emptyList(),
        bcc: List<String> = emptyList(),
    ): Email = suspendCancellableCoroutine { continuation ->
        try {
            gateway.request(
                method = "email.send",
                params = mapOf(
                    "account_id" to accountId,
                    "to" to to,
                    "subject" to subject,
                    "body" to body,
                    "cc" to cc,
                    "bcc" to bcc,
                ),
            ) { response ->
                try {
                    val email = json.decodeFromJsonElement(
                        Email.serializer(),
                        response as JsonElement
                    )
                    continuation.resume(email)
                } catch (e: Exception) {
                    continuation.resumeWithException(
                        EmailError.DecodingError(e.message ?: "Unknown")
                    )
                }
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    suspend fun saveDraft(
        accountId: String,
        to: List<String>,
        subject: String,
        body: String,
        cc: List<String> = emptyList(),
        bcc: List<String> = emptyList(),
    ): EmailDraft = suspendCancellableCoroutine { continuation ->
        try {
            gateway.request(
                method = "email.draft.save",
                params = mapOf(
                    "account_id" to accountId,
                    "to" to to,
                    "subject" to subject,
                    "body" to body,
                    "cc" to cc,
                    "bcc" to bcc,
                ),
            ) { response ->
                try {
                    val draft = json.decodeFromJsonElement(
                        EmailDraft.serializer(),
                        response as JsonElement
                    )
                    continuation.resume(draft)
                } catch (e: Exception) {
                    continuation.resumeWithException(
                        EmailError.DecodingError(e.message ?: "Unknown")
                    )
                }
            }
        } catch (e: Exception) {
            continuation.resumeWithException(e)
        }
    }

    suspend fun getAnalytics(accountId: String): EmailAnalytics =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "email.analytics",
                    params = mapOf("account_id" to accountId),
                ) { response ->
                    try {
                        val analytics = json.decodeFromJsonElement(
                            EmailAnalytics.serializer(),
                            response as JsonElement
                        )
                        continuation.resume(analytics)
                    } catch (e: Exception) {
                        continuation.resumeWithException(
                            EmailError.DecodingError(e.message ?: "Unknown")
                        )
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    suspend fun getAttachmentUrl(emailId: String, attachmentId: String): String =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "email.attachment.url",
                    params = mapOf("email_id" to emailId, "attachment_id" to attachmentId),
                ) { response ->
                    try {
                        val url = (response as? JsonElement)?.jsonObject?.get("url")?.toString()
                            ?: throw EmailError.DecodingError("No URL in response")
                        continuation.resume(url)
                    } catch (e: Exception) {
                        continuation.resumeWithException(e)
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }

    suspend fun getThread(threadId: String): List<Email> =
        suspendCancellableCoroutine { continuation ->
            try {
                gateway.request(
                    method = "email.thread.get",
                    params = mapOf("thread_id" to threadId),
                ) { response ->
                    try {
                        val emails = (response as? JsonElement)?.jsonArray?.map {
                            json.decodeFromJsonElement(Email.serializer(), it)
                        } ?: emptyList()
                        continuation.resume(emails)
                    } catch (e: Exception) {
                        continuation.resumeWithException(
                            EmailError.DecodingError(e.message ?: "Unknown")
                        )
                    }
                }
            } catch (e: Exception) {
                continuation.resumeWithException(e)
            }
        }
}

File: helix-runtime/apps/ios/Sources/Email/EmailService.swift
import Foundation

/// Service for communicating with email functionality via gateway
class EmailService {
    private let gateway: GatewaySession
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    init(gateway: GatewaySession) {
        self.gateway = gateway
        decoder.dateDecodingStrategy = .iso8601
        encoder.dateEncodingStrategy = .iso8601
    }

    // MARK: - Email Fetching

    /// Fetch paginated list of emails
    func fetchEmails(limit: Int = 20, skip: Int = 0) async throws -> (emails: [Email], pagination: EmailPaginationInfo) {
        let response = try await gateway.request(
            "email.list",
            params: [
                "limit": limit,
                "skip": skip
            ]
        )

        let data = try JSONSerialization.data(withJSONObject: response, options: [])
        let result = try decoder.decode(EmailListResponse.self, from: data)

        return (result.emails, result.pagination)
    }

    /// Get a specific email with full content
    func getEmail(_ id: String) async throws -> Email {
        let response = try await gateway.request(
            "email.get",
            params: ["id": id]
        )

        let data = try JSONSerialization.data(withJSONObject: response, options: [])
        return try decoder.decode(Email.self, from: data)
    }

    // MARK: - Email Search

    /// Search emails with filters
    func searchEmails(filter: EmailSearchFilter, limit: Int = 20) async throws -> [Email] {
        var params: [String: Any] = ["limit": limit]

        if !filter.query.isEmpty {
            params["query"] = filter.query
        }
        if let from = filter.from {
            params["from"] = from
        }
        if let to = filter.to {
            params["to"] = to
        }
        if let subject = filter.subject {
            params["subject"] = subject
        }
        if let label = filter.label {
            params["label"] = label
        }
        if let isRead = filter.isRead {
            params["is_read"] = isRead
        }
        if let isStarred = filter.isStarred {
            params["is_starred"] = isStarred
        }
        if let startDate = filter.startDate {
            params["start_date"] = ISO8601DateFormatter().string(from: startDate)
        }
        if let endDate = filter.endDate {
            params["end_date"] = ISO8601DateFormatter().string(from: endDate)
        }
        if let hasAttachments = filter.hasAttachments {
            params["has_attachments"] = hasAttachments
        }

        let response = try await gateway.request("email.search", params: params)

        let data = try JSONSerialization.data(withJSONObject: response, options: [])
        let result = try decoder.decode(EmailSearchResponse.self, from: data)

        return result.results
    }

    // MARK: - Email Actions

    /// Mark email as read or unread
    func markAsRead(_ id: String, isRead: Bool) async throws {
        _ = try await gateway.request(
            "email.mark_read",
            params: [
                "id": id,
                "is_read": isRead
            ]
        )
    }

    /// Mark email as starred or unstarred
    func markAsStarred(_ id: String, isStarred: Bool) async throws {
        _ = try await gateway.request(
            "email.mark_starred",
            params: [
                "id": id,
                "is_starred": isStarred
            ]
        )
    }

    /// Add label to email
    func addLabel(_ emailId: String, label: String) async throws {
        _ = try await gateway.request(
            "email.add_label",
            params: [
                "email_id": emailId,
                "label": label
            ]
        )
    }

    /// Remove label from email
    func removeLabel(_ emailId: String, label: String) async throws {
        _ = try await gateway.request(
            "email.remove_label",
            params: [
                "email_id": emailId,
                "label": label
            ]
        )
    }

    /// Delete email (soft delete)
    func deleteEmail(_ id: String) async throws {
        _ = try await gateway.request(
            "email.delete",
            params: ["id": id]
        )
    }

    /// Permanently delete email
    func deleteEmailPermanently(_ id: String) async throws {
        _ = try await gateway.request(
            "email.delete_permanent",
            params: ["id": id]
        )
    }

    // MARK: - Compose & Send

    /// Send a new email
    func sendEmail(_ draft: EmailDraft) async throws -> Email {
        var params: [String: Any] = [
            "to": draft.to.map { ["email": $0.email, "name": $0.name as Any] },
            "subject": draft.subject,
            "body": draft.body
        ]

        if !draft.cc.isEmpty {
            params["cc"] = draft.cc.map { ["email": $0.email, "name": $0.name as Any] }
        }

        if !draft.bcc.isEmpty {
            params["bcc"] = draft.bcc.map { ["email": $0.email, "name": $0.name as Any] }
        }

        if let inReplyTo = draft.inReplyTo {
            params["in_reply_to"] = inReplyTo
        }

        let response = try await gateway.request("email.send", params: params)

        let data = try JSONSerialization.data(withJSONObject: response, options: [])
        return try decoder.decode(Email.self, from: data)
    }

    /// Save draft
    func saveDraft(_ draft: EmailDraft) async throws -> String {
        var params: [String: Any] = [
            "to": draft.to.map { ["email": $0.email, "name": $0.name as Any] },
            "subject": draft.subject,
            "body": draft.body
        ]

        if !draft.cc.isEmpty {
            params["cc"] = draft.cc.map { ["email": $0.email, "name": $0.name as Any] }
        }

        if !draft.bcc.isEmpty {
            params["bcc"] = draft.bcc.map { ["email": $0.email, "name": $0.name as Any] }
        }

        let response = try await gateway.request("email.save_draft", params: params)

        guard let draftId = response as? String else {
            throw EmailError.decodingError("Invalid draft ID response")
        }

        return draftId
    }

    // MARK: - Analytics

    /// Get email analytics
    func getAnalytics() async throws -> EmailAnalytics {
        let response = try await gateway.request("email.analytics", params: [:])

        let data = try JSONSerialization.data(withJSONObject: response, options: [])
        return try decoder.decode(EmailAnalytics.self, from: data)
    }

    /// Get email statistics for date range
    func getStatistics(startDate: Date, endDate: Date) async throws -> [EmailFrequencyData] {
        let formatter = ISO8601DateFormatter()
        let response = try await gateway.request(
            "email.statistics",
            params: [
                "start_date": formatter.string(from: startDate),
                "end_date": formatter.string(from: endDate)
            ]
        )

        let data = try JSONSerialization.data(withJSONObject: response, options: [])
        let result = try decoder.decode(EmailStatisticsResponse.self, from: data)

        return result.data
    }

    // MARK: - Attachments

    /// Get attachment download URL
    func getAttachmentUrl(_ emailId: String, attachmentId: String) async throws -> URL {
        let response = try await gateway.request(
            "email.attachment_url",
            params: [
                "email_id": emailId,
                "attachment_id": attachmentId
            ]
        )

        guard let urlString = response as? String, let url = URL(string: urlString) else {
            throw EmailError.decodingError("Invalid attachment URL")
        }

        return url
    }

    // MARK: - Thread Operations

    /// Get all emails in a thread
    func getThread(_ threadId: String) async throws -> [Email] {
        let response = try await gateway.request(
            "email.thread",
            params: ["thread_id": threadId]
        )

        let data = try JSONSerialization.data(withJSONObject: response, options: [])
        let result = try decoder.decode(EmailThreadResponse.self, from: data)

        return result.emails
    }
}

// MARK: - Response Types

private struct EmailListResponse: Codable {
    let emails: [Email]
    let pagination: EmailPaginationInfo
}

private struct EmailSearchResponse: Codable {
    let results: [Email]
}

private struct EmailStatisticsResponse: Codable {
    let data: [EmailFrequencyData]
}

private struct EmailThreadResponse: Codable {
    let emails: [Email]
}
